### ISSUE TITLE
Exception throwed when accessing a freed StreamPipeReader ReadResult.Buffer is unhelpful.

### ISSUE THREAD

**Issue by Kuinox** [dotnet/runtime#53794]
### Description

```csharp
byte[] buffer = new byte[42];
PipeReader pipe = PipeReader.Create( new MemoryStream( buffer ) );
ReadResult readResult = await pipe.ReadAsync();
readResult.Buffer.ToArray();
pipe.Complete();
readResult.Buffer.ToArray();//throws.
```
The exception thrown is `System.ArgumentOutOfRangeException : Specified argument was out of the range of valid values. (Parameter 'start')`
Because of some mistake, the Complete happened concurrently, which made it hard to debug and to understand why this error happened.

**Comment by davidfowl** [comment]
I'm not sure how doable that is since the ReadResult is a read only struct without a back pointer to the PipeReader (it's a state snapshot). The way to avoid these issues is to only complete from the code that is doing the reading. I'd suggest reading this thoroughly https://docs.microsoft.com/en-us/dotnet/standard/io/pipelines

Tl;DR Never call Complete as a way to cancel a pending operation and always call Complete from the code consuming the PipeReader or PipeWriter

**Comment by Kuinox** [comment]
> I'd suggest reading this thoroughly https://docs.microsoft.com/en-us/dotnet/standard/io/pipelines  

I did read this doc a lot :D  
But, I don't know why, I didn't found this line previously:  
> The reader and writer loops end by calling Complete. Complete lets the underlying Pipe release the memory it allocated.
